src/plugins/score-plugin-fx/Fx/FactorOracle2.hpp

```diff
@@ -0,0 +1,425 @@
+    \param i The integer of the current state.
+    \param word The input string.
+
+*/
+/*! \fn int LengthCommonSuffix(FactorOracle& States, int phi_one, int phi_two)
+    \brief Finds the length of a common suffix ending at the position phi_one and phi_two by traversing the suffix links.
+    \param phi_one The position of the state.
+    \param phi_two The position of the state.
+*/
+/*! \fn int FindBetter(FactorOracle& States, vector <vector<int>> &T, int i, char alpha, string word)
+    \brief Writes \a count bytes from \a buf to the filedescriptor \a fd.
+    \param T A vector where each position has all the suffix transitions directed to each state.
+    \param i The integer of the current state.
+    \param alpha The transition symbol.
+    \param word The input string.
+    \return A better state
+*/
+/*! \fn string FOGenerate(FactorOracle& States, int i, string v, float q)
+    \brief Generates the Factor Oracle improvisation.
+    \param i The integer of the current state.
+    \param v The sequence v.
+    \param q A float argument.
+    \return The factor oracle improvisation
+*/
+/*! \fn void FactorOracleStart(FactorOracle& OracleRelations,string word)
+    \brief Starts the process of the Factor Oracle generation .
+    \param word The input string.
+*/
+
+namespace Factor
+{
+/** The class SingleTransition denotes the elements that belong to each transition
+  */
+template <class T>
+class SingleTransition
+{
+public:
+  int first_state_; /**< denotes the first state of the transition */
+  int last_state_;  /**< denotes the last state of the transition */
+  T symbol_;        /**< denotes the symbol (letter) of the transition */
+};
+
+/** The class State denotes the elements that belong to each state
+ * state denotes de number of the state
+ * std::vector <SingleTransition> transition is the std::vector where all forward links of the state are defined
+ * suffix_transition denotes which is the suffix link of this state
+ * lrs is the longest repeated subsequence of this state
+ * */
+template <class T>
+class State
+{
+public:
+  int state_; /*!< denotes the number of the state */
+  std::vector<SingleTransition<T>>
+      transition_; /*!< denotes the number of the state */
+  int suffix_transition_;
+  int lrs_ = 0;
+  void singleTransitionResize() { transition_.resize(10); }
+};
+
+template <class T>
+class FactorOracle
+{
+public:
+  int phi, k, fo_iter, current_state = 1;
+  std::vector<T> input_values;
+  std::vector<State<T>> states_; /**< std::vector of all the states */
+  std::vector<std::vector<int>>
+      RevSuffix; /**< std::vector where each position has all the suffix transitions directed to each state */
+  FactorOracle()
+  {
+    this->states_.resize(2);
+    this->states_[0].state_ = 0;
+    this->states_[0].lrs_ = 0;
+    this->states_[0].suffix_transition_ = -1; /*!< S[0] = -1 */
+    this->RevSuffix.resize(2);
+  }
+  void AddLetter(int i, std::vector<T> word)
```

Comment:
Since `word` is only read from it should be passed as const-ref : right now it is being copied every time which is pretty bad for real-time performance as this incurs memory allocations